### PR TITLE
feat: 1368 authorization

### PR DESCRIPTION
**Pull request by ebezzi**
### Reviewers
**Functional:** 
@MDunitz 

**Readability:** 

---

Issue [1368](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/1368)

## Changes
Implements token based authorization. Adds a `requires_auth` and `requires_scope` decorators that can be used to restrict APIs. 
Notes: 
1. The current authentication flow isn't ideal. The proper Oauth2 flow requires the token not to be stored in the cookie, but rather in the FE local storage/memory and sent via the `Authorization` header to each request. The `id_token` is also probably useless for this application and should be replaced with a call to `userinfo`
2. `connexion` (which is currently use to manage secured access to the APIs via the swagger yaml file) does not support scopes unless you use the `Oauth2` flow (currently we're using a custom flow). Therefore, we need to do this manually using decorators. Fixing 1. would allow to do this properly
3. We need two different audiences: one for the `id_token` (currently in configuration) and another for the `access_token` (to be stored in config.audience2 - need to figure out how to do that).

### CODE REVIEW

**Comment by tihuan** [comment]
Oh yes totally! Thanks so much, Emanuele 🤩🚀

**Review by MDunitz — APPROVED** [review]
lgtm

File: backend/corpora/lambdas/api/v1/authentication.py [hunk 1]
```diff
@@ -47,7 +48,8 @@ def get_oauth_client(config: CorporaAuthConfig) -> FlaskRemoteApp:
         refresh_token_url=config.api_token_url,
         access_token_url=config.api_token_url,
         authorize_url=config.api_authorize_url,
-        client_kwargs={"scope": "openid profile email offline_access"},
+        client_kwargs={"scope": "openid profile email offline_access read:collections write:collections"},
```
Comment (MDunitz) [inline backend/corpora/lambdas/api/v1/authentication.py:51]: did we want to get rid of read:collections?

File: oauth/api_scopes.json [hunk 2]
```diff
@@ -1,5 +1,11 @@
 [
   {
     "Name": "test_scope"
+  },
+  {
+    "Name": "read:collections"
```
Comment (MDunitz) [inline oauth/api_scopes.json:6]: drop?

File: oauth/clients-config.json [hunk 3]
```diff
@@ -12,7 +12,7 @@
             "http://backend.corporanet.local:5000/dp/v1/oauth2/callback"
         ],
         "AllowedScopes": [
-           "openid", "profile", "email", "offline_access"
+           "openid", "profile", "email", "offline_access", "read:collections", "write:collections"
```
Comment (MDunitz) [inline oauth/clients-config.json:15]: drop read:collections?

File: backend/corpora/lambdas/api/v1/authentication.py (hunk 1 above)
Comment (ebezzi) [reply backend/corpora/lambdas/api/v1/authentication.py:51]: I was thinking to leave it for future implementations, but we can always add it back later. Will remove it altogether.

**Pull request merged**